subsys/net/lib/dns/dns_cache.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2024 Endress+Hauser AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/dns_resolve.h>
+#include "dns_cache.h"
+
+LOG_MODULE_REGISTER(net_dns_cache, CONFIG_DNS_RESOLVER_LOG_LEVEL);
+
+static void dns_cache_clean(struct dns_cache const *cache);
+
+int dns_cache_flush(struct dns_cache *cache)
+{
+	k_mutex_lock(cache->lock, K_FOREVER);
+	for (size_t i = 0; i < cache->size; i++) {
+		cache->entries[i].in_use = false;
+	}
+	k_mutex_unlock(cache->lock);
+
+	return 0;
+}
+
+int dns_cache_add(struct dns_cache *cache, char const *query, struct dns_addrinfo const *addrinfo,
+		  uint32_t ttl)
+{
+	int64_t oldest_entry_uptime = INT64_MAX;
+	size_t index_to_replace = -1;
+	bool found_empty = false;
+
+	if (cache == NULL || query == NULL || addrinfo == NULL || ttl == 0) {
+		return -EINVAL;
+	}
+
+	if (strlen(query) >= CONFIG_DNS_RESOLVER_MAX_QUERY_LEN) {
+		NET_WARN("Query string to big to be processed %u >= "
+			 "CONFIG_DNS_RESOLVER_MAX_QUERY_LEN",
+			 strlen(query));
+		return -EINVAL;
+	}
+
+	k_mutex_lock(cache->lock, K_FOREVER);
+
+	NET_DBG("ADD \"%s\" with TTL %" PRIu32, query, ttl);
+
+	for (size_t i = 0; i < cache->size; i++) {
+		if (!cache->entries[i].in_use) {
+			index_to_replace = i;
+			found_empty = true;
+		}
+	}
+
+	if (!found_empty) {
+		dns_cache_clean(cache);
```

Comment:
Maybe just clean the cache before the first loop, so that we don't need to check for `in_use` twice?

Reply:
Thats what I had before. Reverted to stage where this was the case.